internal/locate/region_cache.go

```diff
@@ -1203,6 +1204,160 @@ func (c *RegionCache) LocateKeyRange(bo *retry.Backoffer, startKey, endKey []byt
+		opt.needBuckets = true
+	}
+}
+
+func WithNeedRegionHasLeaderPeer() BatchLocateKeyRangesOpt {
+	return func(opt *batchLocateKeyRangesOption) {
+		opt.needRegionHasLeaderPeer = true
+	}
+}
+
+// BatchLocateKeyRanges lists regions in the given ranges.
+func (c *RegionCache) BatchLocateKeyRanges(bo *retry.Backoffer, keyRanges []kv.KeyRange, opts ...BatchLocateKeyRangesOpt) ([]*KeyLocation, error) {
+	uncachedRanges := make([]pd.KeyRange, 0, len(keyRanges))
+	cachedRegions := make([]*Region, 0, len(keyRanges))
+	// 1. find regions from cache
+	for _, keyRange := range keyRanges {
+		for {
+			r := c.tryFindRegionByKey(keyRange.StartKey, false)
+			if r == nil {
+				// region cache miss, add the cut range to uncachedRanges, load from PD later.
+				uncachedRanges = append(uncachedRanges, pd.KeyRange{StartKey: keyRange.StartKey, EndKey: keyRange.EndKey})
+				break
+			}
+			// region cache hit, add the region to cachedRegions.
+			cachedRegions = append(cachedRegions, r)
+			if r.ContainsByEnd(keyRange.EndKey) {
+				// the range is fully hit in the region cache.
+				break
+			}
+			keyRange.StartKey = r.EndKey()
+		}
+	}
+
+	merger := newBatchLocateRegionMerger(cachedRegions, len(cachedRegions)+len(uncachedRanges))
+
+	// 2. load remaining regions from pd client
+	for len(uncachedRanges) > 0 {
+		regions, err := c.BatchLoadRegionsWithKeyRanges(bo, uncachedRanges, defaultRegionsPerBatch, opts...)
+		if err != nil {
+			return nil, err
+		}
+		if len(regions) == 0 {
+			return nil, errors.Errorf("BatchLoadRegionsWithKeyRanges return empty batchedRegions without err")
+		}
+
+		for _, r := range regions {
+			merger.appendRegion(r)
+		}
+		// if the regions are not loaded completely, split uncachedRanges and load the rest.
+		uncachedRanges = nextRanges(uncachedRanges, regions[len(regions)-1].EndKey())
+	}
+	return merger.build(), nil
+}
+
+type batchLocateRangesMerger struct {
+	lastEndKey      util.Option[[]byte]
+	cachedIdx       int
+	cachedRegions   []*Region
+	mergedLocations []*KeyLocation
+}
+
+func newBatchLocateRegionMerger(cachedRegions []*Region, sizeHint int) *batchLocateRangesMerger {
+	return &batchLocateRangesMerger{
+		lastEndKey:      util.None[[]byte](),
+		cachedRegions:   cachedRegions,
+		mergedLocations: make([]*KeyLocation, 0, sizeHint),
+	}
+}
+
+func (m *batchLocateRangesMerger) appendKeyLocation(r *Region) {
+	m.mergedLocations = append(m.mergedLocations, &KeyLocation{
+		Region:   r.VerID(),
+		StartKey: r.StartKey(),
+		EndKey:   r.EndKey(),
+		Buckets:  r.getStore().buckets,
+	})
+}
+
+func (m *batchLocateRangesMerger) appendRegion(uncachedRegion *Region) {
+	defer func() {
+		endKey := uncachedRegion.EndKey()
+		if len(endKey) == 0 {
+			m.cachedIdx = len(m.cachedRegions)
+		} else {
+			m.lastEndKey = util.Some(uncachedRegion.EndKey())
+		}
+	}()
+	if len(uncachedRegion.StartKey()) == 0 {
+		m.appendKeyLocation(uncachedRegion)
+		return
+	}
+	lastEndKey := m.lastEndKey.Inner()
```

Comment:
Could the `util.Option` be avoided? The check is still whether the pointer is not nil.